lib/dialyzer/src/dialyzer_typegraph.erl

```diff
@@ -0,0 +1,105 @@
+
+%% The module type deps of a module are modules that depend on the module to
+%% define types, contracts, callbacks or behaviours
+module_type_deps(UseContracts, CodeServer, Modules) ->
+
+  Contracts =
+    case UseContracts of
+      true -> maps:from_list(dict:to_list(dialyzer_codeserver:get_contracts(CodeServer)));
+      false -> []
+    end,
+  Callbacks = maps:from_list(dialyzer_codeserver:get_callbacks(CodeServer)),
+  TypeDefinitions =
+    maps:from_list(
+      [{M, dialyzer_codeserver:lookup_mod_records(M, CodeServer)} || M <- Modules]
+    ),
+  Behaviours =
+    maps:from_list(
+      [{M, get_behaviours_for_module(M, CodeServer)} || M <- Modules]
+    ),
+  collect_module_type_deps(Contracts, Callbacks, TypeDefinitions, Behaviours).
+
+-spec get_behaviours_for_module(module(), dialyzer_codeserver:codeserver()) -> [module()].
+
+get_behaviours_for_module(M, CodeServer) ->
+  ModCode = dialyzer_codeserver:lookup_mod_code(M, CodeServer),
+  Attrs = cerl:module_attrs(ModCode),
+  {Behaviours, _BehaviourLocations} = dialyzer_behaviours:get_behaviours(Attrs),
+  Behaviours.
+
+-spec collect_module_type_deps(Specs, Callbacks, TypeDefinitions, Behaviours) -> type_mod_deps() when
+    Specs :: #{mfa() => dialyzer_contracts:file_contract()},
+    Callbacks :: #{mfa() => dialyzer_contracts:file_contract()},
+    TypeDefinitions :: #{module() => erl_types:type_table()},
+    Behaviours :: #{module() => [module()]}.
+
+collect_module_type_deps(Specs, Callbacks, TypeDefinitions, Behaviours) ->
+
+  Contracts =
+    [{M, Spec} || {{M, _F, _A}, {_FileLine, Spec, _Extra}} <- maps:to_list(Specs)] ++
+    [{M, Callback} || {{M, _F, _A}, {_FileLine, Callback, _Extra}} <- maps:to_list(Callbacks)],
+
+  ModulesMentionedInTypeDefinitions =
+    [{FromTypeDefM, erl_types:module_type_deps_of_type_defs(TypeTable)} || {FromTypeDefM, TypeTable} <- maps:to_list(TypeDefinitions)],
+
+  ModulesMentionedInContracts =
+    [{FromContractM, module_type_deps_of_contract(C)} || {FromContractM, C} <- Contracts],
+
+  ModulesMentionedAsBehaviours =
+    [{M, gb_sets:from_list(BehaviourList)} || {M, BehaviourList} <- maps:to_list(Behaviours)],
+
+  AllDepsRaw =
+    ModulesMentionedInContracts ++
+    ModulesMentionedInTypeDefinitions ++
+    ModulesMentionedAsBehaviours,
+
+  %% Find the union of module dependencies from all sources, removing self-references
+  AllDeps =
+    lists:foldl(
+      fun ({M,Deps}, Acc) ->
+        NewDepsSetWithoutSelf = gb_sets:delete_any(M, Deps),
+        maps:update_with(
+          M,
+          fun (ExistingDepsSet) ->
+            gb_sets:union(NewDepsSetWithoutSelf, ExistingDepsSet)
+          end,
+          NewDepsSetWithoutSelf,
+          Acc)
+      end,
+      maps:new(),
+      AllDepsRaw),
+
+  ModuleToWhatItDependsOn = maps:map(fun (_Mod, Deps) -> gb_sets:to_list(Deps) end, AllDeps),
+
+  ModuleToThoseModulesThatDependOnIt = dialyzer_utils:transpose_map_list(ModuleToWhatItDependsOn),
+
```

Comment:
The `sofs` module can be a little intimidating, so here is an example how you can use `sofs`:

```suggestion
  AllDepsRaw1 = [{M, gb_sets:to_list(Deps)} || {M, Deps} <- AllDepsRaw],
  S0 = sofs:relation(AllDepsRaw1, [{atom,[atom]}]),
  S1 = sofs:relation_to_family(S0),
  S2 = sofs:family_union(S1),
  S3 = sofs:family_to_relation(S2),
  S4 = sofs:converse(S3),
  S5 = sofs:relation_to_family(S4),
  S6 = [{M, Ds -- [M]} || {M, Ds} <- sofs:to_external(S5)],
  ModuleToThoseModulesThatDependOnIt = maps:from_list(S6),

```

(I leave it to you to avoid creating the `gb_sets` in the first place, so that you can eliminate the list comprehension in first line.)